lib/chef/knife/bootstrap.rb

```diff
@@ -246,7 +247,19 @@ def default_bootstrap_template
       #
       # @return [String] The DNS or IP that bootstrap will connect to
       def server_name
-        Array(@name_args).first
+        host_descriptor[:host]
+      end
+
+      def host_descriptor
+        unless @name_args.nil?
+          parts = Array(@name_args).first.split('@')
+          @host_descriptor = {
+            :host => parts.pop,
+            :user => parts.pop
+          }
+        end
+
+        @host_descriptor
```

Comment:
This is probably better written something like this:

```ruby
def first_arg
  Array(@name_args).first
end

def server_name
   @server_name ||= if first_arg.include?('@')
      first_arg.split('@')[1]
    else
      first_arg
    end
end

def user_name
  @user_name ||=
    first_arg.split('@')[0] if first_arg.include?('@')
  # this will be nil in the else case
end
```

i don't quite see where you've handled the case when there is no '@' in this code either.

also not sure you need to check `@name_args.nil?` since that would cause the existing `server_name` method to explode I don't think that's a useful case to worry about and some other code must handle it.